lib/src/iam/jwks.rs

```diff
@@ -0,0 +1,686 @@
+#[cfg(not(test))]
+static CACHE_COOLDOWN: Lazy<chrono::Duration> =
+	Lazy::new(|| match std::env::var("SURREAL_JWKS_CACHE_COOLDOWN") {
+		Ok(seconds_str) => {
+			let seconds = seconds_str
+				.parse::<u64>()
+				.expect("Expected a valid number of seconds for SURREAL_JWKS_CACHE_COOLDOWN");
+			Duration::seconds(seconds as i64)
+		}
+		Err(_) => {
+			Duration::seconds(300) // Set default cache refresh cooldown of 5 minutes
+		}
+	});
+
+static REMOTE_TIMEOUT: Lazy<chrono::Duration> =
+	Lazy::new(|| match std::env::var("SURREAL_JWKS_REMOTE_TIMEOUT") {
+		Ok(milliseconds_str) => {
+			let milliseconds = milliseconds_str
+				.parse::<u64>()
+				.expect("Expected a valid number of milliseconds for SURREAL_JWKS_REMOTE_TIMEOUT");
+			Duration::milliseconds(milliseconds as i64)
+		}
+		Err(_) => {
+			Duration::milliseconds(1000) // Set default remote timeout to 1 second
+		}
+	});
+
+// Generates a verification configuration from a JWKS object hosted in a remote location
+// Performs local caching of all JWKS objects to prevent unnecessary network requests
+// Implements checks to prevent denial of service and unauthorized network requests
+// Validates the JWK objects found in the JWKS object according to RFC 7517
+// Source: https://datatracker.ietf.org/doc/html/rfc7517
+pub(super) async fn config(
+	kvs: &Datastore,
+	kid: &str,
+	url: &str,
+) -> Result<(DecodingKey, Validation), Error> {
+	// Attempt to fetch relevant JWK object either from local cache or remote location
+	let jwk = match fetch_jwks_from_cache(url).await {
+		Ok(jwks_cache) => {
+			trace!("Successfully fetched JWKS object from local cache");
+			// Check that the cached JWKS object has not expired yet
+			if Utc::now().signed_duration_since(jwks_cache.time) < *CACHE_EXPIRATION {
+				// Attempt to find JWK in JWKS object from local cache
+				match jwks_cache.jwks.find(kid) {
+					Some(jwk) => jwk.to_owned(),
+					_ => {
+						trace!("Could not find valid JWK object with key identifier '{kid}' in cached JWKS object");
+						// Check that the cached JWKS object has not been recently updated
+						if Utc::now().signed_duration_since(jwks_cache.time) < *CACHE_COOLDOWN {
+							debug!("Refused to refresh cache before cooldown period is over");
+							return Err(Error::InvalidAuth); // Return opaque error
+						}
+						find_jwk_from_url(kvs, url, kid).await?
+					}
+				}
+			} else {
+				trace!("Fetched JWKS object from local cache has expired");
+				find_jwk_from_url(kvs, url, kid).await?
+			}
+		}
+		Err(_) => {
+			trace!("Could not fetch JWKS object from local cache");
+			find_jwk_from_url(kvs, url, kid).await?
+		}
+	};
+
+	// Check if algorithm specified is supported
+	let alg = match jwk.common.algorithm {
+		Some(alg) => alg,
+		_ => {
+			warn!("Invalid value for parameter 'alg' in JWK object: '{:?}'", jwk.common.algorithm);
+			return Err(Error::InvalidAuth); // Return opaque error
+		}
+	};
```

Comment:
Just a style note you can handle these types of branches with a let else statement:
```rust
let Some(alg) = jwk.common.algorithm else {
    warn!("Invalid value for parameter 'alg' in JWK object: '{:?}'", jwk.common.algorithm")
    return Err(Error::InvalidAuth);
};
```
I personally prefer them because there a bit shorter, bit this way is also fine.

Reply:
Thank you! I was not aware of this syntax, it definitely seems cleaner for many cases. In this instance, we want to keep the value of `alg` to use it at the end of the `config` function. I can't see a way to do it without having both the `if` and the `else` clauses, which would end up similar to the `match` statement. Is that correct?

Reply:
Oh, I had no idea this is how it worked. I agree than it would have been more elegant then, although probably not worth it to change it now considering that the tests take 15 minutes to run and often require a re-run and we are wanting to have this merged ASAP :sweat:

Reply:
Agreed, it was just a style note, it isn't something that really needs to be changed.